### PR TITLE
Automatically materialize CTEs

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR adds functionality to inspect query plans before binding, to find out whether CTEs should be materialized. Currently, DuckDB never materializes CTEs. If CTEs are queried multiple times, they are recomputed, which can be very expensive for large complex CTEs.

The heuristic introduced in this PR is pretty simple: if the CTE performs a (grouped) aggregation and is queried more than once, it should be materialized. This heuristic can be improved in a later PR to also materialize when, e.g., the CTE contains a window function.

Pipeline sharing is something that is on our roadmap for the future, which serves a similar purpose while being more generally useful, but this is much more difficult to implement, so that will have to wait.

To make sure query plans do not regress, I've improved CTE statistics in the join order optimizer.

Some performance improvement highlights below.

TPC-H
```
benchmark/tpch/sf1/q15.benchmark
Old timing: 0.046487
New timing: 0.024154
```

TPC-DS
```
benchmark/tpcds/sf1/q23.benchmark
Old timing: 0.655913
New timing: 0.288761

benchmark/tpcds/sf1/q24.benchmark
Old timing: 0.053736
New timing: 0.031949

benchmark/tpcds/sf1/q36.benchmark
Old timing: 0.141676
New timing: 0.061156

benchmark/tpcds/sf1/q57.benchmark
Old timing: 0.173906
New timing: 0.089459

benchmark/tpcds/sf1/q59.benchmark
Old timing: 0.187678
New timing: 0.097933

benchmark/tpcds/sf1/q64.benchmark
Old timing: 0.240837
New timing: 0.158492
```

### CODE REVIEW

**Comment by kryonix** [comment]
Hi @lnkuiper cool idea! I've briefly discussed a similar (yet non existing) PR with @Mytherin a while back.

I just glimpsed over the code, so not sure if this is actually an issue currently. But I think this optimization should not be applied when a CTE explicitly requests to be `NOT MATERIALIZED`, aka. when `CTEMaterialize::CTE_MATERIALIZE_NEVER` is set.

**Review by Tmonster — APPROVED** [review]
LGTM w.r.t. the join order optimizer. Had a couple of comments about some of the other logic

Also don't quite understand the reference counting stuff, but I think that's just to get a count of the CTEs in a query plan right?

File: src/optimizer/join_order/relation_manager.cpp [hunk 1]
```diff
@@ -263,7 +264,35 @@ bool RelationManager::ExtractJoinRelations(LogicalOperator &input_op,
...
+		auto &cte_ref = op->Cast<LogicalCTERef>();
+		if (cte_ref.materialized_cte != CTEMaterialize::CTE_MATERIALIZE_ALWAYS) {
+			return false;
+		}
+		AddRelation(input_op, parent, optimizer.GetMaterializedCTEStats(cte_ref.cte_index));
+		return true;
+	}
 	default:
+		// non-reorderable operator, just recurse into children
+		for (auto &child : op->children) {
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply src/optimizer/join_order/relation_manager.cpp:292]: You're right, thanks!

File: src/optimizer/join_order/relation_manager.cpp [hunk 2]
```diff
@@ -263,7 +264,35 @@ bool RelationManager::ExtractJoinRelations(LogicalOperator &input_op,
...
+		auto lhs_stats = RelationStats();
+		// optimize the lhs child and copy the stats
+		auto lhs_optimizer = optimizer.CreateChildOptimizer();
+		op->children[0] = lhs_optimizer.Optimize(std::move(op->children[0]), &lhs_stats);
+		// optimize the rhs child
+		auto rhs_optimizer = optimizer.CreateChildOptimizer();
+		auto table_index = op->type == LogicalOperatorType::LOGICAL_MATERIALIZED_CTE
+		                       ? op->Cast<LogicalMaterializedCTE>().table_index
+		                       : op->Cast<LogicalRecursiveCTE>().table_index;
+		rhs_optimizer.AddMaterializedCTEStats(table_index, std::move(lhs_stats));
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply src/optimizer/join_order/relation_manager.cpp:278]: The LHS can be a `LogicalMaterializedCTE` or a `LogicalRecursiveCTE` that is materialized. In both cases we encounter a `LogicalCTERef`, so I had to add the case for `LogicalRecursiveCTE` to get proper join plans

File: src/planner/binder.cpp [hunk 3]
```diff
@@ -205,6 +206,87 @@ void Binder::AddCTEMap(CommonTableExpressionMap &cte_map) {
...
+			continue;
+		}
+
+		// if the cte is a SELECT node
+		if (cte.second->query->node->type != QueryNodeType::SELECT_NODE) {
+			continue;
+		}
+
+		// we materialize if the CTE ends in an aggregation
+		auto &cte_node = cte.second->query->node->Cast<SelectNode>();
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply src/planner/binder.cpp:273]: Indeed, if the CTE is more complex, we wouldn't materialize it. This PR implements a basic heuristic for materialization, which could be expanded in a later PR.

We need to be careful with materializing joins: In the worst case, we have a non-selective join between two large tables. If we materialize that, we might make the query use much more memory and degrade the performance. For aggregations, it's much simpler, as aggregations are already a materialization, so we don't really risk degrading performance by putting the result in a materialized CTE.

**Comment by lnkuiper** [comment]
I've also added statistics to `LOGICAL_DELIM_GET` in the join order optimizer, so we can now reorder join plans with delim gets in them. This slightly improves TPC-H Q2

**Comment by lnkuiper** [comment]
Overall, TPC-DS improved, but I couldn't stand Q31 and Q75 regressing.

They regressed because the CTEs queried more than once had very selective filters on them, which were pushed down. This made executing the CTE multiple times (with different filters) faster than materializing the entire CTE once and then filtering. I've added an optimizer that `OR`s the filters on top of CTE references and pushes them into the materialized CTE definition, now we no longer regress.

**Comment by Tmonster** [comment]
Do you think it's possible to add some micro benchmarks? These changes affect tpcds performance and we don't benchmark tpcds regularly with our CI. We do, however, run the micro-benchmarks every week, so by adding a micro benchmark we can keep a better eye on regresions

**Comment by lnkuiper** [comment]
@Tmonster I can add micro-benchmark, but [TPC-DS regression tests are run on every push](https://github.com/duckdb/duckdb/actions/runs/9397448096/job/25880683058?pr=12290).

**Comment by Tmonster** [comment]
Right, for some reason I thought it was only tpch. I would still advocate for a benchmark, because the tpcds queries are fairly complex and if a query regresses, it might be for reasons unrelated to materialized CTEs. 
A micro benchmark could address this by specifically testing materializing CTEs and limited other functionality.
I just feel like while the TPCH and TPCDS improvements are great, we should not rely on our benchmarks to point out that a specific functionality has regressed

**Comment by lnkuiper** [comment]
I've added `benchmark/tpch/cte/auto_cte_materialization.benchmark`, which is 1.5-2x faster compared to `feature`.

**Comment by Mytherin** [comment]
Thanks! LGTM - could you just solve the merge conflict?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**